### DISCUSSION TITLE
How does dropout work and when should I use it?

### DISCUSSION THREAD

**Discussion by aasimansari1** [aasimansari1/ml-interview-prep#22] [Q&A] [answered]
How does dropout work and when should I use it?

**Comment by aasimansari1** [comment, marked as answer]
**Dropout** randomly deactivates neurons during training — forcing the network to learn redundant representations, which reduces overfitting.

## How it works:
```python
import torch.nn as nn

# During TRAINING: each neuron set to 0 with probability p
# During INFERENCE: all neurons active, outputs scaled by (1-p)

dropout = nn.Dropout(p=0.5)  # 50% dropout rate

class Network(nn.Module):
    def __init__(self):
        super().__init__()
        self.fc1 = nn.Linear(512, 256)
        self.drop1 = nn.Dropout(0.5)   # after hidden layer
        self.fc2 = nn.Linear(256, 128)
        self.drop2 = nn.Dropout(0.3)   # lighter dropout near output
        self.fc3 = nn.Linear(128, 10)
    
    def forward(self, x):
        x = self.drop1(self.fc1(x).relu())
        x = self.drop2(self.fc2(x).relu())
        return self.fc3(x)  # NO dropout on output layer!
```

## Critical: train vs eval mode
```python
model.train()   # dropout ACTIVE — use during training
model.eval()    # dropout DISABLED — use during inference/validation

# Easy mistake to miss:
with torch.no_grad():
    model.eval()           # ← must call this!
    preds = model(X_test)
```

## When to use:
- ✅ Large fully-connected layers (p=0.3-0.5)
- ✅ When overfitting is a problem (train acc >> val acc)
- ❌ Small models — dropout reduces capacity too much
- ❌ After BatchNorm (they conflict) — use one or the other
- ❌ Output layer — always keep active

## Modern alternatives:
- **DropPath** (Stochastic Depth): drops entire residual branches — better for transformers
- **Variational dropout**: same mask each forward pass — better for RNNs
